beetsplug/replaygain.py

```diff
@@ -308,42 +307,47 @@ def compute_album_gain(self, task):
         target_level_lufs = db_to_lufs(task.target_level)
 
         # analyse tracks
-        # list of track Gain objects
-        track_gains = []
-        # maximum peak
-        album_peak = 0
-        # sum of BS.1770 gating block powers
-        sum_powers = 0
-        # total number of BS.1770 gating blocks
-        n_blocks = 0
-
-        for item in task.items:
-            track_gain, track_n_blocks = self._analyse_item(
-                item, task.target_level, task.peak_method
+        # Gives a list of tuples (track_gain, track_n_blocks)
+        track_results = [
+            self._analyse_item(
+                item,
+                task.target_level,
+                task.peak_method,
+                count_blocks=True,
             )
-            track_gains.append(track_gain)
+            for item in task.items
+        ]
+
+        # list of track Gain objects
+        track_gains = [tg for tg, _nb in track_results]
 
-            # album peak is maximum track peak
-            album_peak = max(album_peak, track_gain.peak)
+        # Album peak is maximum track peak
+        album_peak = max(tg.peak for tg in track_gains)
 
-            # prepare album_gain calculation
-            # total number of blocks is sum of track blocks
-            n_blocks += track_n_blocks
+        # Total number of BS.1770 gating blocks
+        n_blocks = sum(nb for _tg, nb in track_results)
 
+        def sum_of_track_powers(track_gain, track_n_blocks):
             # convert `LU to target_level` -> LUFS
-            track_loudness = target_level_lufs - track_gain.gain
+            loudness = target_level_lufs - track_gain
```

Comment:
This line seems broken now; after track gains are computed the album gain errors out with
```
TypeError: unsupported operand type(s) for -: 'int' and 'Gain'
```

Looks like it should have been called with `sum_of_track_powers(tg.gain, nb)` below.

Reply:
Ouch, thanks for testing this thoroughly (which I obviously didn't)!

What happens here is that that our audio test files contain only silence, which the final result that the tests don't actually hit this calculation (more detail: the number of gating blocks with loudness of above the BS1770 gating threshold `nblocks` is zero, even though at ~1s the track is in principle longer that a single gating block of 400ms).

@sampsyo it looks like you added the test files (`full.mp3`, etc). Is there a script around that you used to generate them? Maybe we could replace them with a simple sine wave or so, instead of just silence.

Reply:
Good question—I just generated those by taking a short silent WAV and using various tools (ffmpeg? maybe even iTunes, back in the day?) to manually convert it to different formats. It's a good idea to do something similar with an off-the-shelf one-second sine wave or similar…